Add driver simulator

nicolasiensen opened the pull request:
This is a first version of the driver simulator.

Inside of the driver-simulator folder, you will find:
- a `README` file to explain how to use the simulator;
- a `simulate.js` which implements the simulator;
- a `vehicles.json` used to simulate the vehicles

The simulator is capable of register a vehicle, update its location and de-register the vehicle when it reaches its destination.

Improvements to this solution would include a docker configuration in order to simplify the setup and add more vehicles to the JSON file.

Jira ticket: RSD-605

---
- [x] Fix the duplicated `lat` attribute in the update of the vehicle location in the simulator
- [x] Add a `.nvmrc` file to have a default node version in the project
- [x] Add tests to functions listed in the API module
- [x] Handle connection errors with the API
- [ ] ~Test how the simulator performs with a big amount of vehicles (100)~

nicolasiensen committed:
Add driver simulator

This is a first version of the driver simulator.

Inside of the driver-simulator folder you will find:
- a REDME file to explain how to use the simulator;
- a simulate.js which implements the simulator;
- a vehicles.json used to simulate the vehicles

The simulator is capable of register a vehicle, update its location and
de-register the vehicle when it reaches its destination.

Improvements to this solution would include a docker configuration in
order to simplify the setup, and add more vehicles with new routes to
the JSON file.

adamlukemooney commented on driver-simulator/README.md:
```diff
@@ -0,0 +1,18 @@
+
+## Get started
+
+Install [node](https://github.com/creationix/nvm#install-script) and then run the following command:
```
It would be good to specify the desired Node version, and perhaps include a `.nvmrc`.

nicolasiensen replied:
👍

adamlukemooney commented on driver-simulator/simulate.js:
```diff
@@ -0,0 +1,55 @@
+  request(
+    `/vehicles/${vehicleId}/locations`,
+    "POST",
+    { lat: location[1], lat: location[0], at: new Date().toISOString() },
```
You have two `lat` values here.

It should be: `{ lat: location[0], lng: location[1], /* ... */ }`

nicolasiensen replied:
Good catch!

adamlukemooney reviewed (changes requested):
Nice job!

I have highlighted a couple of points in the review.

Also, I'd like to see:
- a couple of small tests
- handling of connection errors (e.g. the candidate's solution is not running at the address they specified)

We should also check how well the simulation performs when there are 100 vehicles running. I think it should be fine, as we're just talking to localhost, but it might slow down a bit with remote hosts, and then we might want to consider a multi-threaded solution. Let's see how well it works.

nicolasiensen committed:
Split the simulator from the API interface

Jira ticket: RSD-605

nicolasiensen committed:
Fix the update of a vehicle position

Fix the request body to update the vehicle position and simplify the API
interface.

Jira ticket: RSD-605

nicolasiensen committed:
Add the nvmrc file

Jira ticket: RSD-605

nicolasiensen committed:
Add jest for testing

Since now we have dependencies to manage (jest) we are also introducing
in this commit a yarn.lock file.

Also, new scripts where created in the package.json file in order to
start the simulator, and run the tests using jest.

Jira ticket: RSD-605

nicolasiensen committed:
Add tests for the api module and document how to run them

Jira ticket: RSD-605

nicolasiensen committed:
Add tests for the simulation

Jira ticket: RSD-605

nicolasiensen committed:
Handle connection errors

Jira ticket: RSD-605

nicolasiensen commented:
OMG, I've learned so much about jest doing this simulator!

nicolasiensen commented:
I'm going to leave the task `Test how the simulator performs with a big amount of vehicles (100)` for another pull request, ok?
